### ISSUE TITLE
Define how ending bracket should be placed

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#355]
# Rule request

## Thesis

Currently one can write both:

```python
some(
    1,
    2,
    3)
```

and

```python
some(
    1,
    2,
    3,
)
```

I suggest to stick to the second example for consistency. 

I am not sure that this will be possible without an external plugin.

## Reasoning

That's mostly a consistency thing. This is just a preference.

**Comment by sobolevn** [comment]
This one is very hard. That's what I came up with:

## Violation

```python
@final
class WrongBracketPositionViolation(TokenizeViolation):
    """
    Forbid to use extra indentation.

    Reasoning:
        You can use extra indentation for lines of code.
        Python allows you to do that in case you will keep the indentation
        level equal for this specific node.
        But, that's insane!

    Solution:
        Place bracket on the same line, when a single line expression.
        Or place bracket on a new line when a multi-line expression.

    Example::

        # Correct:
        print([
            1, 2, 3,
        ])

        print(
            1,
            2,
        )

        def _annotate_brackets(
            tokens: List[tokenize.TokenInfo],
        ) -> TokenLines:
            ...

        # Wrong:
        print([
            1, 2, 3],
        )

        print(
            1,
            2)

        def _annotate_brackets(
            tokens: List[tokenize.TokenInfo]) -> TokenLines:
            ...

    We check round, square, and curly brackets.

    .. versionadded:: 0.6.0

    """

    should_use_text = False
    error_template = 'Found bracket in wrong position'
    code = 319

```

## Solution

```python
@final
class BracketLocationVisitor(BaseTokenVisitor):
    """
    Finds closing brackets location.

    We check that brackets can be on the same line or
    brackets can be the only tokens on the line.

    We track all kind of brackets: round, square, and curly.
    """

    # TODO: I am pretty sure we can tweak performance of this class.
    # Try set() instead of tuple, optimize _post_visit(), etc
    _opening_tokens: ClassVar[Tuple[int, ...]] = (
        tokenize.LBRACE,
        tokenize.LSQB,
        tokenize.LPAR,
    )

    _closing_tokens: ClassVar[Tuple[int, ...]] = (
        tokenize.RBRACE,
        tokenize.RSQB,
        tokenize.RPAR,
    )

    _allowed_empty_line_tokens: ClassVar[Tuple[int, ...]] = (
        tokenize.NL,
        tokenize.NEWLINE,
        *_closing_tokens,
    )

    def __init__(self, *args, **kwargs) -> None:
        """Creates line tracking for tokens."""
        super().__init__(*args, **kwargs)
        self._lines: TokenLines = defaultdict(list)

    def _annotate_brackets(
        self,
        tokens: List[tokenize.TokenInfo],
    ) -> TokenLines:
        """Annotates each opening bracket with the nested level index."""
        annotated: TokenLines = defaultdict(list)
        level = 0
        for token in tokens:
            if token.exact_type in self._opening_tokens:
                level += 1
                annotated[level].append(token)
            if token.exact_type in self._closing_tokens:
                if annotated[level]:
                    annotated[level].pop()
                level -= 1
        return annotated

    def _is_single_line(self, tokens: List[tokenize.TokenInfo]) -> bool:
        if not contains(tokens, self._opening_tokens):
            return False

        annotated_lines = self._annotate_brackets(tokens)
        for annotated in annotated_lines.values():
            if annotated:
                return False
        return True

    def _is_annotation(self, tokens: List[tokenize.TokenInfo]) -> bool:
        for token in tokens:
            if token.string == '->':
                return True
        return False

    def _check_contents(self, tokens: List[tokenize.TokenInfo]) -> bool:
        if not contains(tokens, self._closing_tokens):
            return True
        if only_contains(tokens, self._allowed_empty_line_tokens):
            return True
        if self._is_single_line(tokens) or self._is_annotation(tokens):
            return True
        return False

    def _post_visit(self) -> None:
        for _line, tokens in self._lines.items():
            if not tokens:
                continue
            if self._check_contents(tokens):
                continue

            self.add_violation(WrongBracketPositionViolation(tokens[-1]))

    def visit(self, token: tokenize.TokenInfo) -> None:
        """Goes trough all tokens to separate them by lines."""
        self._lines[token.start[0]].append(token)

```

But, it does not solve several problems:
1. It just ignores lines with `->` completely, it does not check anything in the annotations
2. It has bugs, I have check this plugin on the linter itself. It reports tons of false positive errors.

I guess we need another algorithm.

How current version works?
1. It ignores lines with closing brackets
2. It ignores lines that only contains closing brackets
3. It ignores lines when closing brackets are on the same line as the opening one: `print()`
4. It ignores annotations (broken)

How that should be done?

```python
def function(
    arg: int,
) -> Tuple[int, 'SomeClass']:
     ...
```

1. We should check that annotation line starts with `)` exactly
2. We should check that annotation fits into one line (we may update `flake8-type-annotations` for that or create a separate rule for that (I am leaning to a separate rule)
3. We should check that line ends with `:\n`
